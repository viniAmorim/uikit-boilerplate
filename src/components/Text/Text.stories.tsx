/* eslint-disable no-alert */
import React from 'react';

import { Story, Meta } from '@storybook/react';

import Text, { ParagraphProps } from './Text';

export default {
  title: 'Atoms/Typography/Text',
  component: Text,
} as Meta;

const Template: Story<ParagraphProps> = (args) => <Text {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Ad astra per aspera',
};

export const FontColor = Template.bind({});
FontColor.args = {
  children: 'Ad astra per aspera',
  fontColor: 'pink500',
};

export const Weight = Template.bind({});
Weight.args = {
  children: 'Ad astra per aspera',
  weight: '300',
};

export const FontSize = Template.bind({});
FontSize.args = {
  children: 'Ad astra per aspera',
  fontSize: '16px',
};

export const LineHeight = Template.bind({});
LineHeight.args = {
  children: 'Ad astra per aspera',
  lineHeight: '26px',
};

export const Link = Template.bind({});
Link.args = {
  children: 'Ad astra per aspera',
  onClick: () => alert('Click!'),
};
