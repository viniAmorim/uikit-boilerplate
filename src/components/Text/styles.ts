import styled, { css } from 'styled-components';

import { ColorOptions } from '../types';

export type StyledParagraphProps = {
  weight: string;
  fontSize: string;
  fontColor: ColorOptions;
  lineHeight: string;
};

export const Paragraph = styled.div<StyledParagraphProps>`
  font-family: 'Montserrat';
  font-size: ${(props) => props.fontSize};
  font-weight: ${(props) => props.weight};
  line-height: ${(props) => props.lineHeight};
  color: ${({ theme, fontColor = 'black700' }) => theme.colors[fontColor]};

  ${({ onClick }) =>
    onClick &&
    css`
      cursor: pointer;
    `}
`;
