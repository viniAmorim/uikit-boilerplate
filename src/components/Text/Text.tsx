import React, { BaseHTMLAttributes, forwardRef, ReactNode } from 'react';

import * as Styled from './styles';

export interface ParagraphProps
  extends BaseHTMLAttributes<HTMLParagraphElement>,
    Partial<Styled.StyledParagraphProps> {
  children: ReactNode;
  onClick?: () => void;
}

const Text = forwardRef<HTMLParagraphElement, ParagraphProps>(function Text(
  {
    children,
    weight = '700',
    fontSize = '24px',
    fontColor = 'black700',
    lineHeight = '34px',
    onClick,
    ...rest
  },
  ref,
) {
  return (
    <Styled.Paragraph
      weight={weight}
      fontSize={fontSize}
      fontColor={fontColor}
      lineHeight={lineHeight}
      onClick={onClick}
      ref={ref}
      {...rest}
    >
      {children}
    </Styled.Paragraph>
  );
});

export default Text;
