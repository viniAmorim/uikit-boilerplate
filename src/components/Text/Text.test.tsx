import React from 'react';

import snapshotDiff from 'snapshot-diff';

import { fireEvent, render, waitFor } from '~/test/utils';

import Text from './Text';

describe('Text', () => {
  it('should render correctly', () => {
    const { getByText } = render(<Text>Text</Text>);

    const textElement = getByText('Text');

    expect(textElement).toBeInTheDocument();
    expect(textElement).toHaveTextContent(/^Text$/);
    expect(textElement).toMatchSnapshot();
  });

  it('should render different FontColors', () => {
    const { rerender, asFragment } = render(
      <Text fontColor="pink500">Text</Text>,
    );
    const pink = asFragment();

    rerender(<Text fontColor="grey500">Text</Text>);
    const grey = asFragment();

    rerender(<Text fontColor="purple700">Text</Text>);
    const purple = asFragment();

    expect(snapshotDiff(pink, grey)).toMatchSnapshot();
    expect(snapshotDiff(pink, purple)).toMatchSnapshot();
    expect(snapshotDiff(grey, purple)).toMatchSnapshot();
  });

  it('should render different Weights', () => {
    const { rerender, asFragment } = render(<Text weight="300">Text</Text>);
    const first = asFragment();

    rerender(<Text weight="400">Text</Text>);
    const second = asFragment();

    rerender(<Text weight="700">Text</Text>);
    const third = asFragment();

    expect(snapshotDiff(first, second)).toMatchSnapshot();
    expect(snapshotDiff(first, third)).toMatchSnapshot();
    expect(snapshotDiff(second, third)).toMatchSnapshot();
  });

  it('should render different FontSizes', () => {
    const { rerender, asFragment } = render(<Text fontSize="14px">Text</Text>);
    const first = asFragment();

    rerender(<Text fontSize="16px">Text</Text>);
    const second = asFragment();

    rerender(<Text fontSize="24px">Text</Text>);
    const third = asFragment();

    expect(snapshotDiff(first, second)).toMatchSnapshot();
    expect(snapshotDiff(first, third)).toMatchSnapshot();
    expect(snapshotDiff(second, third)).toMatchSnapshot();
  });

  it('should render different LineHeight', () => {
    const { rerender, asFragment } = render(
      <Text lineHeight="24px">Text</Text>,
    );
    const first = asFragment();

    rerender(<Text lineHeight="26px">Text</Text>);
    const second = asFragment();

    rerender(<Text lineHeight="34px">Text</Text>);
    const third = asFragment();

    expect(snapshotDiff(first, second)).toMatchSnapshot();
    expect(snapshotDiff(first, third)).toMatchSnapshot();
    expect(snapshotDiff(second, third)).toMatchSnapshot();
  });

  it('should have onClick', async () => {
    const handleClick = jest.fn();

    const { getByText } = render(<Text onClick={handleClick}>Text</Text>);

    const textElement = getByText('Text');

    fireEvent.click(textElement);

    await waitFor(() => {
      expect(handleClick).toHaveBeenCalledTimes(1);
      expect(handleClick).toHaveBeenCalled();
      expect(textElement).toHaveStyle('cursor: pointer');
    });
  });
});
