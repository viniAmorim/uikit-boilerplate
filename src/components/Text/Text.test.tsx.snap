// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render correctly 1`] = `
<div
  class="sc-beqWaB fpoiqw"
  font-size="24px"
>
  Text
</div>
`;

exports[`Text should render different FontColors 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB bdgNmx\\"
+     class=\\"sc-beqWaB irBlpR\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different FontColors 2`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB bdgNmx\\"
+     class=\\"sc-beqWaB jPoJot\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different FontColors 3`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB irBlpR\\"
+     class=\\"sc-beqWaB jPoJot\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different FontSizes 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB dQkonX\\"
-     font-size=\\"14px\\"
+     class=\\"sc-beqWaB cHoNiN\\"
+     font-size=\\"16px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different FontSizes 2`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB dQkonX\\"
-     font-size=\\"14px\\"
+     class=\\"sc-beqWaB fpoiqw\\"
+     font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different FontSizes 3`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB cHoNiN\\"
-     font-size=\\"16px\\"
+     class=\\"sc-beqWaB fpoiqw\\"
+     font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different LineHeight 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB hLWyKh\\"
+     class=\\"sc-beqWaB gDLPiL\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different LineHeight 2`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB hLWyKh\\"
+     class=\\"sc-beqWaB fpoiqw\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different LineHeight 3`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB gDLPiL\\"
+     class=\\"sc-beqWaB fpoiqw\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different Weights 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB iYNEkQ\\"
+     class=\\"sc-beqWaB iNNjkD\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different Weights 2`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB iYNEkQ\\"
+     class=\\"sc-beqWaB fpoiqw\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;

exports[`Text should render different Weights 3`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
-     class=\\"sc-beqWaB iNNjkD\\"
+     class=\\"sc-beqWaB fpoiqw\\"
      font-size=\\"24px\\"
    >
      Text
    </div>
  </DocumentFragment>"
`;
